### PR TITLE
Updating `CMakeLists.txt`

### PR DESCRIPTION
**Pull request by landinjm**
Updating the `CMakeLists.txt` to be more understandable and modular.

I've adapted the one used by [ASPECT](https://github.com/geodynamics/aspect/tree/main) which is much more helpful in terms of error handling. For example, it will warn if the dependencies are not installed or if there is a leftover `CMakeCache.txt` 

This PR also adds [Catch2 ](https://github.com/catchorg/Catch2/tree/devel)for unit testing in `catch.hpp`

### CODE REVIEW

**Comment by landinjm** [comment]
Note that this does change the behavior in how executable files are generated. Before, you would run make release or make debug to generate `main.exe`. Now, running make generates `main-release.exe` and `main-debug.exe`. The user can then choose which one to use.

File: .gitignore
Comment (fractalsbyx) [inline .gitignore:1]: add newline at end

Comment (landinjm) [reply .gitignore:1]: Done

**Review by fractalsbyx — APPROVED** [review]
wow, so much better!

**Pull request merged**